storey/targets.py

```diff
@@ -888,48 +891,100 @@ def _init(self):
-                    value = record.get(self._columns[column_index], "NULL")
-                    b.write(self._sanitize_value(value))
-                    if column_index < len(self._columns) - 1:
-                        b.write(",")
-                b.write(") ")
-            b.write(";")
-            insert_statement = b.getvalue()
-        self._connection.execute(insert_statement)
+                sql.write(f" USING {self._supertable} TAGS {self._get_params_template(self._number_of_tags)}")
+            sql.write(f" VALUES {self._get_params_template(self._number_of_values)};")
+            return sql.getvalue()
+
+    @classmethod
+    def _get_unsupported_error(cls, value) -> UnsupportedTDEngineTypeError:
+        return cls.UnsupportedTDEngineTypeError(f"Unsupported value type {type(value)} in {cls.__name__}")
+
+    @classmethod
+    def _value_to_tag(cls, value):
+        import taosws
+
+        if isinstance(value, bool):
+            return taosws.bool_to_tag(value)
+        if isinstance(value, int):
+            return taosws.int_to_tag(value)
+        if isinstance(value, float):
+            return taosws.float_to_tag(value)
+        if isinstance(value, str):
+            return taosws.varchar_to_tag(value)
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
+        return tags
+
+    @classmethod
+    def _raw_value_to_column_fun(cls, value) -> Callable:
+        import taosws
+
+        if isinstance(value, datetime.datetime):
+            return taosws.millis_timestamps_to_column
+        if isinstance(value, bool):
+            return taosws.bools_to_column
+        if isinstance(value, int):
+            return taosws.ints_to_column
+        if isinstance(value, float):
+            return taosws.floats_to_column
+        if isinstance(value, str):
+            return taosws.nchar_to_column
+        raise cls._get_unsupported_error(value)
+
+    @classmethod
+    def _raw_value_to_value(cls, value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)
+        return value
+
+    def _get_batch_values(self, batch: list[dict]) -> list:
+        values = [[] for _ in range(self._number_of_values)]
+        to_column_funs = []
+        val_names = self._columns[self._number_of_tags :]
+        for event_ind, event in enumerate(batch):
+            for val_ind, val_name in enumerate(val_names):
+                raw_value = event.get(val_name)
+                if event_ind == 0:
+                    to_column_funs.append(self._raw_value_to_column_fun(raw_value))
+                values[val_ind].append(self._raw_value_to_value(raw_value))
+        return [fun(vals) for fun, vals in zip(to_column_funs, values)]
```

Comment:
Seems unnecessarily functional to me. I'd do something like
```python
    def _get_batch_values(self, batch: list[dict]) -> list:
        result = []
        val_names = self._columns[self._number_of_tags :]
        for event in batch:
            values = []
            for val_name in val_names:
                raw_value = event.get(val_name)
                values.append(self._raw_value_to_value(raw_value))
            to_column_func = self._raw_value_to_column_fun(event.get(val_names[0]))
            result.append(to_column_func(values))
        return result
```
(potentially dealing with `event.get(val_names[0])` differently, and do note that I didn't test this snippet)